### PR TITLE
operator: fast-fail if the leader changed to the RemovePeer

### PR DESCRIPTION
**Pull request by nolouch**
Signed-off-by: nolouch <nolouch@gmail.com>

### What problem does this PR solve?
Fix https://github.com/pingcap/pd/issues/2493
If the leader changed cause by TiKV(maybe too busy or network problem), the operator cannot step to the final state, such as
```
PD:
try to remove peer 24
TiKV:
reject remove peer 24 because it is a leader.
``` 
Then the operator needs 10mins to timeout, which may cause the rule sync learner became slow.

### Check List

Tests

- Unit test

### Release note

- Let the operator fast fail if the leader changed to the RemovePeer

### CODE REVIEW

File: server/schedule/operator_controller.go [hunk 1]
```diff
@@ -147,7 +147,26 @@ func (oc *OperatorController) Dispatch(region *core.RegionInfo, source string) {
 	}
 }
 
-func (oc *OperatorController) checkStaleOperator(op *operator.Operator, region *core.RegionInfo) bool {
+func (oc *OperatorController) checkStaleOperator(op *operator.Operator, step operator.OpStep, region *core.RegionInfo) bool {
+	switch s := step.(type) {
+	case operator.RemovePeer:
+		if s.FromStore == region.GetLeader().GetStoreId() {
+			if oc.removeOperatorWithoutBury(op) {
```
Comment (rleungx) [inline server/schedule/operator_controller.go:154]: How about using a new function for line 154 to line 167

**Comment by nolouch** [comment]
PTAL @rleungx @HunDunDM

File: server/schedule/operator_controller.go [hunk 2]
```diff
@@ -147,7 +147,21 @@ func (oc *OperatorController) Dispatch(region *core.RegionInfo, source string) {
 	}
 }
 
-func (oc *OperatorController) checkStaleOperator(op *operator.Operator, region *core.RegionInfo) bool {
+func (oc *OperatorController) checkStaleOperator(op *operator.Operator, step operator.OpStep, region *core.RegionInfo) bool {
+	switch s := step.(type) {
+	case operator.RemovePeer:
+		if s.FromStore == region.GetLeader().GetStoreId() {
```
Comment (disksing) [inline server/schedule/operator_controller.go:153]: I think it's too tricky to make special judgments on removePeer and leader here. Maybe we can solve it by recording the expected region leader at each step and check if the leader is the same. (like `op.ConfVerChanged`)

File: server/schedule/operator_controller.go [hunk 3]
```diff
@@ -147,7 +147,21 @@ func (oc *OperatorController) Dispatch(region *core.RegionInfo, source string) {
...
+			if oc.RemoveOperator(op,
+				zap.String("reason", "stale operator, cannot remove leader per"),
```
Comment (rleungx) [inline server/schedule/operator_controller.go:155]: typo?

File: server/schedule/operator_controller.go [hunk 4]
```diff
@@ -552,6 +562,17 @@ func (oc *OperatorController) buryOperator(op *operator.Operator) {
...
 			zap.Reflect("operator", op))
 		operatorCounter.WithLabelValues(op.Desc(), "timeout").Inc()
+	case operator.CANCELED:
+		fileds := []zap.Field{
+			zap.Uint64("region-id", op.RegionID()),
+			zap.Duration("takes", op.RunningTime()),
+			zap.Reflect("operator", op),
+		}
+		fileds = append(fileds, extraFileds...)
+		log.Info("operator caceled",
```
Comment (lhy1024) [inline server/schedule/operator_controller.go:572]:
```suggestion
		log.Info("operator cancelled",
```

File: server/schedule/operator/step.go [hunk 5]
```diff
@@ -130,6 +153,21 @@ func (al AddLearner) IsFinish(region *core.RegionInfo) bool {
...
+func (al AddLearner) CheckSafety(region *core.RegionInfo) error {
+	peer := region.GetStorePeer(al.ToStore)
+	if peer == nil {
+		return nil
+	}
+	if peer.GetId() != al.PeerID {
+		return errors.Errorf("peer id is not matched, expect %d, got %d", al.PeerID, peer.GetId())
+	}
+	if peer.IsLearner {
+		return errors.New("peer already is a voter")
```
Comment (rleungx) [inline server/schedule/operator/step.go:166]: learner?

File: server/schedule/operator/step.go [hunk 6]
```diff
@@ -101,6 +115,15 @@ func (ap AddPeer) Influence(opInfluence OpInfluence, region *core.RegionInfo) {
 	to.AdjustStepCost(storelimit.RegionAdd, regionSize)
 }
 
+// CheckSafety checks if the step meets the safety.
+func (ap AddPeer) CheckSafety(region *core.RegionInfo) error {
+	peer := region.GetStorePeer(ap.ToStore)
+	if peer != nil && peer.GetId() != ap.PeerID {
+		return errors.Errorf("peer id is not matched, expect %d, got %d", ap.PeerID, peer.GetId())
```
Comment (rleungx) [inline server/schedule/operator/step.go:122]: How about
```suggestion
		return errors.Errorf("peer %d has already existed in store %d, the operator is trying to add peer %d on the same store", peer.GetId(), ToStore, ap.PeerID)
```

File: server/schedule/operator/step.go [hunk 7]
```diff
@@ -52,6 +54,18 @@ func (tl TransferLeader) IsFinish(region *core.RegionInfo) bool {
 	return region.GetLeader().GetStoreId() == tl.ToStore
 }
 
+// CheckSafety checks if the step meets the safety.
```
Comment (disksing) [inline server/schedule/operator/step.go:57]: meets safy problem?

**Comment by nolouch** [comment]
PTAL @disksing @rleungx

**Pull request merged**